File: enode/src/main/java/org/enodeframework/eventing/impl/InMemoryPublishedVersionStore.java
package org.enodeframework.eventing.impl;

import org.enodeframework.eventing.IPublishedVersionStore;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

/**
 * @author anruence@gmail.com
 */
public class InMemoryPublishedVersionStore implements IPublishedVersionStore {

    private final CompletableFuture<Void> successTask = CompletableFuture.completedFuture(null);

    private final ConcurrentMap<String, Integer> versionDict = new ConcurrentHashMap<>();

    @Override
    public CompletableFuture<Void> updatePublishedVersionAsync(String processorName, String aggregateRootTypeName, String aggregateRootId, int publishedVersion) {
        versionDict.put(buildKey(processorName, aggregateRootId), publishedVersion);
        return successTask;
    }

    @Override
    public CompletableFuture<Integer> getPublishedVersionAsync(String processorName, String aggregateRootTypeName, String aggregateRootId) {
        int publishedVersion = versionDict.getOrDefault(buildKey(processorName, aggregateRootId), 0);
        return CompletableFuture.completedFuture(publishedVersion);
    }

    private String buildKey(String eventProcessorName, String aggregateRootId) {
        return String.format("%s-%s", eventProcessorName, aggregateRootId);
    }
}

File: enode/src/main/java/org/enodeframework/eventing/impl/InMemoryPublishedVersionStore.kt
package org.enodeframework.eventing.impl

import org.enodeframework.eventing.PublishedVersionStore
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap

/**
 * @author anruence@gmail.com
 */
class InMemoryPublishedVersionStore : PublishedVersionStore {
    private val versionDict: ConcurrentMap<String, Int> = ConcurrentHashMap()
    override fun updatePublishedVersionAsync(
        processorName: String,
        aggregateRootTypeName: String,
        aggregateRootId: String,
        publishedVersion: Int
    ): CompletableFuture<Int> {
        versionDict[buildKey(processorName, aggregateRootId)] = publishedVersion
        return CompletableFuture.completedFuture(1)
    }

    override fun getPublishedVersionAsync(
        processorName: String,
        aggregateRootTypeName: String,
        aggregateRootId: String
    ): CompletableFuture<Int> {
        val publishedVersion = versionDict.getOrDefault(buildKey(processorName, aggregateRootId), 0)
        return CompletableFuture.completedFuture(publishedVersion)
    }

    private fun buildKey(eventProcessorName: String, aggregateRootId: String): String {
        return "$eventProcessorName-$aggregateRootId"
    }
}